## Importing the libraries 

In [4]:
import pandas as pd
import numpy as np

## Load the data

In [18]:
path_to_file = 'workshop_orders.csv'  

df = pd.read_csv(path_to_file)
# print(df)
grouped = {}
# print(df[1,1])
print(df["customer_id"][1])
for ind in df.index:
    customer = df["customer_id"][ind]
    if customer in grouped:
        grouped[customer].append((df["order_status"][ind], df["order_amount"][ind], df["order_created_at"][ind]))
    else:
        grouped[customer] = [(df["order_status"][ind], df["order_amount"][ind], df["order_created_at"][ind])]
print(grouped.keys)
for key in grouped:
    print(f"Key is {key} and value is {grouped[key]}")


b1bbe434-6d4e-46e5-9e54-6e785e0ad047
<built-in method keys of dict object at 0x12d688780>
Key is 9d1adcfb-288b-42a2-93c0-3e6ab8df2205 and value is [('approved', 0.99, '2022-08-08 02:00:50'), ('declined', 29.99, '2022-09-10 02:02:47'), ('approved', 29.99, '2022-08-11 02:01:47')]
Key is b1bbe434-6d4e-46e5-9e54-6e785e0ad047 and value is [('declined', 29.99, '2022-09-13 12:55:49'), ('declined', 29.99, '2022-08-01 12:56:02'), ('declined', 29.99, '2022-08-01 12:55:48'), ('declined', 29.99, '2022-09-13 12:56:00'), ('approved', 0.99, '2022-07-29 12:55:28'), ('approved', 29.99, '2022-08-05 12:55:47'), ('declined', 29.99, '2022-08-02 12:55:38'), ('declined', 29.99, '2022-09-14 12:55:39')]
Key is 30ede893-594e-4fc1-9c5c-e28689af78d8 and value is [('declined', 29.99, '2022-07-30 16:22:02'), ('declined', 29.99, '2022-07-30 16:21:48'), ('declined', 29.99, '2022-08-14 16:21:42'), ('declined', 29.99, '2022-07-31 16:21:39'), ('declined', 29.99, '2022-08-05 16:21:42'), ('approved', 0.99, '2022-07-27 16:

## EDA

In [3]:
df.head()

In [4]:
df.info()

In [5]:
df["order_amount"].value_counts()

In [6]:
df["order_status"].value_counts()

In [7]:
df.isna().sum()

In [8]:
set(df.duplicated())

## Data transformation

Delete declined transactions.


For convenience, we will transform our data into single-user level (not single order).

## Calculating metrics

In [10]:
res_metrics = {
    "CR_from_trial": 0,
    "Refund_rate": 0,
    "Rebill_rate": 0,
    "Commisions": 0
}

CR from trial = number of converted to full charge users / total amount of users

In [11]:
res_metrics['CR_from_trial'] = # ...
res_metrics['CR_from_trial']

Refund rate = amount of refunded / total earned

In [12]:
res_metrics['Refund_rate'] = # ...
res_metrics['Refund_rate']

Rebill rate:
Rebill rate measures the number of times the average user will pay us for his recurring subscription. To calculate this metric over a 6-month horizon let’s assume that it will decrease each month in 0.7 times starting from the 2nd recurrent payment.

In [13]:
res_metrics['Rebill_rate'] = # ...
res_metrics['Rebill_rate']

And add the commission (10%)

In [14]:
res_metrics['Commisions'] = 0.1

In [15]:
res_metrics

## Calculating LTV
(0.99 + 29.99 * AA % * (1- BB %) * С.СС) * (1 - ZZ %) with AA % being the conversion rate, BB % - the refund rate, C.CC - the rebill rate and ZZ % - the commissions.

In [16]:
LTV = # ...